src/utils/prim-lexing.fs

```diff
@@ -2,10 +2,105 @@
 
 #nowarn "47" // recursive initialization of LexBuffer
 
+namespace Microsoft.FSharp.Compiler.Text
+
+open System
+open System.IO
+
+type ISourceText =
+
+    abstract Item : int -> char with get
+
+    abstract GetLineString : lineIndex: int -> string
+
+    abstract GetLineCount : unit -> int
+
+    abstract GetLastCharacterPosition : unit -> int * int
+
+    abstract GetSubTextString : start: int * length: int -> string
+
+    abstract SubTextEquals : target: string * startIndex: int -> bool
+
+    abstract Length : int
+
+    abstract ContentEquals : sourceText: ISourceText -> bool
+
+    abstract CopyTo : sourceIndex: int * destination: char [] * destinationIndex: int * count: int -> unit  
+
+[<Sealed>]
+type StringText(str: string) =
+
+    let getLines (str: string) =
+        use reader = new StringReader(str)
+        [|
+        let line = ref (reader.ReadLine())
+        while not (isNull !line) do
+            yield !line
+            line := reader.ReadLine()
+        if str.EndsWith("\n", StringComparison.Ordinal) then
+            // last trailing space not returned
+            // http://stackoverflow.com/questions/19365404/stringreader-omits-trailing-linebreak
+            yield String.Empty
+        |]
+
+    let getLines =
+        // This requires allocating and getting all the lines.
+        // However, likely whoever is calling it is using a different implementation of ISourceText
+        // So, it's ok that we do this for now.
+        lazy getLines str
+
+    member __.String = str
+
+    interface ISourceText with
+    
+        member __.Item with get index = str.[index]
+
+        member __.GetLastCharacterPosition() =
+            let lines = getLines.Value
+            if lines.Length > 0 then
+                (lines.Length, lines.[lines.Length - 1].Length)
+            else
+                (0, 0)
+
+        member __.GetLineString(lineIndex) = 
+            getLines.Value.[lineIndex]
+
+        member __.GetLineCount() = getLines.Value.Length
+
+        member __.GetSubTextString(start, length) = 
+            str.Substring(start, length)
+
+        member __.SubTextEquals(target, startIndex) =
+            if startIndex < 0 || startIndex >= str.Length then
+                raise (ArgumentOutOfRangeException("startIndex"))
+
+            if String.IsNullOrEmpty(target) then
+                raise (ArgumentException("Target is null or empty.", "target"))
```

Comment:
Nit: prefer invalidArg as per [guidelines](https://docs.microsoft.com/en-us/dotnet/fsharp/style-guide/conventions#use-exceptions-when-errors-cannot-be-represented-with-types)